### PR TITLE
upgrade rules_go to work with bazel 0.6.0

### LINKED ISSUE #1365: build broken with bazel 0.6.0
I've gotten build errors as follows:

```
% bazel build //...
ERROR: /usr/local/google/home/mukai/.cache/bazel/_bazel_mukai/23d88ec2f9449b2c9d846b080b7cc444/external/io_bazel_rules_go/go/private/binary.bzl:226:48: name 'lib' is not defined (did you mean 'libs'?)
ERROR: error loading package '': Extension 'go/private/binary.bzl' has errors
ERROR: error loading package '': Extension 'go/private/binary.bzl' has errors
INFO: Elapsed time: 0.132s
FAILED: Build did NOT complete successfully (0 packages loaded)
```

Seems this is because of bazel 0.6.0?

I had another environment which still stays on bazel 0.5.4. Succeeded to build, but started reporting the same error after upgraded to 0.6.0.

### PR DESCRIPTION
**Pull request by jmuk**
fixes #1365 

- WORKSPACE; some dependencies are changed slightly, which forces this change. Especially, `go_repositories()` will import protobuf and x/net libraries if not yet imported, so their imports need to come earlier than go_repositories.
- `library` parameter in `go_library` rule seems gone, adapter and template codegen need to be fixed because of this
- fix test/testenv since otherwise build fails.

**Release note**:

```release-note
none
```

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/1366)

### CODE REVIEW

**Comment by kyessenov** [comment]
let's wait until after the release to change go tooling.

**Comment by jmuk** [comment]
sure

**Comment by jmuk** [comment]
presubmit error also says that this (and latest rules_go) isn't compatible with the bazel version used by prow. sigh...

File: WORKSPACE [hunk 1]
```diff
@@ -1,14 +1,36 @@
 workspace(name = "com_github_istio_mixer")
 
 git_repository(
+    name = "org_pubref_rules_protobuf",
```
Comment (douglas-reid) [inline WORKSPACE]: what is the need to pull this out of `adapter_author_deps.bzl`? Can we change the ordering in a different way, so as to support external repo imports better?

File: WORKSPACE [hunk 2]
```diff
@@ -1,14 +1,36 @@
...
+    commit = "0ce5738cd67925351c44df0845c3bbf9d1d32663",  # Sept 22, 2017 (genfiles path calculation fix)
+    remote = "https://github.com/pubref/rules_protobuf",
+)
+
+load("@org_pubref_rules_protobuf//cpp:rules.bzl", "cpp_proto_repositories")
+
+cpp_proto_repositories()
```
Comment (douglas-reid) [inline WORKSPACE]: what is using `cpp_proto_repositories()` within Mixer now?

**Comment by jmuk** [comment]
Let's go with #1423 . Closing.

**Pull request closed without merging**